torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -983,13 +986,25 @@ namespace {
 
 class AsyncBarrierWork : public ProcessGroupGloo::AsyncWork {
  public:
-  AsyncBarrierWork(const std::shared_ptr<gloo::Context>& context, uint32_t tag)
-      : context(context), tag(tag) {}
+  AsyncBarrierWork(
+      const std::shared_ptr<gloo::Context>& context,
+      std::vector<std::weak_ptr<AsyncWork>> priorWork,
```

Comment:
I mean, a RVal ref.

Reply:
Such that passing the parameter in won't incur a copy

Reply:
The current approach also doesn't do any copies. If we were to apply this rule and change all "sink" constructors/functions to take an rvalue then we should do a bigger pass and change everything, not just here. Let's leave this as is here and have a chat on the approach in general (also not just here but elsewhere in PyTorch as well).

Reply:
For myself, I will always do that to make things more explicit, considering that here what you really want to do is to move instead of copy.  Will leave this up to you. I do agree that leaving it this will give people the flexibility to pass either LVal which will copy or RVal which will directly be moved.